core/ResManager/plKeyCollector.cpp

```diff
@@ -141,21 +142,58 @@ void plKeyCollector::cleanupKeys()
 
 void plKeyCollector::reserveKeySpace(const plLocation& loc, short type, int num)
 {
-    keys[loc][type].reserve(num);
+    keys[loc][type]->reserve(num);
+}
+
+uint32_t plKeyCollector::getFlags(const plLocation& loc, short type)
+{
+    return keys[loc][type].getFlags();
+}
+
+void plKeyCollector::setFlags(const plLocation& loc, short type, uint32_t flags)
+{
+    keys[loc][type].setFlags(flags);
 }
 
 void plKeyCollector::sortKeys(const plLocation& loc)
 {
     std::vector<short> types = getTypes(loc);
     for (short type : types) {
-        std::sort(keys[loc][type].begin(), keys[loc][type].end(),
+        std::sort(keys[loc][type]->begin(), keys[loc][type]->end(),
             [](const plKey& a, const plKey& b) { return a->getID() < b->getID(); });
-        for (size_t i = 0; i < keys[loc][type].size(); ++i) {
-            keys[loc][type][i]->setID(i + 1);
+        for (size_t i = 0; i < keys[loc][type]->size(); ++i) {
+            size_t newID = i + 1;
+            if (keys[loc][type][i]->getID() != newID)
+                keys[loc][type][i]->setID(newID);
+
+            // Side effect: verify that pages are telling the truth about
+            // whether or not they are optimized. Fixes incorrect pages
+            // exported by previous revisions of HSPlasma.
+            if (!keys[loc][type].checkFlag(kNotOptimized) && newID < keys[loc][type]->size()) {
+                const plKey& a = keys[loc][type][i];
+                const plKey& b = keys[loc][type][newID];
```

Comment:
Because of these lines, I may have a slight preference for using `i + 1` directly instead of storing it in `newID`...  Or maybe call it `nextID` so it's more clear that it comes from the same sequence as `i` rather than some other random index.